File: ArtOfIllusion/src/main/java/artofillusion/LayoutFileMenu.java
/* Copyright 2024-2026 by Maksim Khramov

   This program is free software; you can redistribute it and/or modify it under the
   terms of the GNU General Public License as published by the Free Software
   Foundation; either version 2 of the License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
   PARTICULAR PURPOSE.  See the GNU General Public License for more details. */


package artofillusion;

import artofillusion.ui.Translate;

import lombok.extern.slf4j.Slf4j;
import org.greenrobot.eventbus.Subscribe;

import javax.swing.*;
import java.util.Objects;

@Slf4j
public final class LayoutFileMenu extends LayoutMenu {

    LayoutFileMenu(LayoutWindow layout) {
        super(layout, "menu.file");

        var mi = Translate.menuItem("new", event -> {
            layout.savePreferences();
            SwingUtilities.invokeLater(ArtOfIllusion::newWindow);
        });
        mi.setIcon(new ImageIcon(Objects.requireNonNull(getClass().getResource("/artofillusion/layout-icons/icons8-new-16.png"))));
        this.add(mi);
        this.add(Translate.menuItem("open", event -> {
            layout.savePreferences();
            SwingUtilities.invokeLater(() -> ArtOfIllusion.openScene(layout));
        }));
    }

    @Subscribe
    public void onSceneChangedEvent(SceneChangedEvent event) {
        log.info("On Scene changed {} ", this.getLayout() == event.window() ? "this" : "other");
    }
}

File: ArtOfIllusion/src/main/kotlin/artofillusion/LayoutFileMenu.kt
package artofillusion

import artofillusion.ui.Translate
import buoy.widget.BMenu
import buoy.widget.BMenuItem
import org.greenrobot.eventbus.Subscribe
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import javax.swing.AbstractAction
import javax.swing.SwingUtilities

class LayoutFileMenu(private val layout: LayoutWindow) : BMenu(Translate.text("menu.file")) {

    init {
        org.greenrobot.eventbus.EventBus.getDefault().register(this)
        this.add(Translate.menuItem("new") { layout.savePreferences(); edt { ArtOfIllusion.newWindow() } })
        this.add(Translate.menuItem("open") { layout.savePreferences(); edt { ArtOfIllusion.openScene(layout) } })
        //Insert recent files here?
        this.add(Translate.menuItem("close") { layout.savePreferences(); edt { ArtOfIllusion.closeWindow(layout) } })
        this.addSeparator()
        this.add(Translate.menuItem("quit") { layout.savePreferences(); edt { ArtOfIllusion.quit() } })
        this.addSeparator()
    }

    @Subscribe
    fun onSceneChangedEvent( event: SceneChangedEvent ): Unit {
        log.info("On Scene changed {} ", if(layout == event.window) "this" else "other")
    }

    private fun edt(action: () -> Unit) = SwingUtilities.invokeLater() { action() }

    companion object  {
        private val log: Logger = LoggerFactory.getLogger(LayoutFileMenu::class.java)
    }

    inner class ImportAction(private val translator: Translator): AbstractAction() {
        override fun actionPerformed(e: java.awt.event.ActionEvent?) {
            translator.importFile(layout)
        }
    }

    inner class ImportSceneMenuItem(private val translator: Translator): BMenuItem(translator.name) {
        init {
            component.addActionListener { translator.importFile(layout) }
        }
    }

    inner class ExportAction(private val translator: Translator): AbstractAction() {
        override fun actionPerformed(e: java.awt.event.ActionEvent?) {
            translator.exportFile(layout, layout.theScene)
        }
    }

    inner class ExportSceneMenuItem(private val translator: Translator): BMenuItem(translator.name) {
        init {
            component.addActionListener { translator.exportFile(layout, layout.theScene) }
        }
    }
}
